### DISCUSSION TITLE
How to move from localStorage persistence to idb-keyval storage

### DISCUSSION THREAD

**Discussion by dohomi** [pmndrs/zustand#2084] [Q&A]
Hello
How can I move a storage from localStorage (previous 1.0 version) to `idb-keyval` storage? If I move without migration the storage is empty after the app starts. But I haven't found the right way on moving the data from localStorage to IndexDB
Thanks for any hints

**Comment by dai-shi** [comment]
You should probably develop your own storage and support migration.
https://github.com/pmndrs/zustand/blob/8e8156f1426105e75f4e041c92bb4561e7dd1e61/src/middleware/persist.ts#L7-L11

**Comment by dohomi** [comment]
Thanks for your input, this is my storage:
```ts
import {del, get, set} from "idb-keyval";
import {StateStorage} from "zustand/middleware";
export const IdbStorage: StateStorage = {
    getItem: async (name) => {
        // Exit early on server
        if (typeof indexedDB === "undefined") {
            return null;
        }
        const value = await get(name);
        console.log("load indexeddb called");
        return value || null;
    },
    setItem: async (name, value) => {
        // Exit early on server
        if (typeof indexedDB === "undefined") {
            return;
        }
        return set(name, value);
    },
    removeItem: async (name) => {
        // Exit early on server
        if (typeof indexedDB === "undefined") {
            return;
        }
        await del(name);
    }
};
```
I have trouble that even I provide a migration the data remains empty after reloading the app:
```ts
    migrate: (prevState, prevVersion) => {
        console.log('migrate', prevState, prevVersion)
        if (prevVersion === 1.0) {
            const oldStorage = localStorage.getItem('tss-storage')
            console.log('oldStorage', oldStorage)
            return oldStorage ? JSON.parse(oldStorage) : prevState
        }
        // return prevState
        return prevState
    }
```

**Reply by dai-shi** [reply]
The `migrate` option is for migrating within the same storage.
Your case is migrating from localStorage to idb, so migration should be done in `storage`.
In other words, don't let Zustand notice that the storage is chagned.

**Reply by dohomi** [reply]
Where is the right place to write the migration? The `StateStorage` interface does not seem to provide a migration, am I overlooking something obvious? Thanks for your swift reply

**Reply by dai-shi** [reply]
I mean it's the app responsibility. You just need to provide the StateStorage interface and implement migration within it. So, basically it's out of the scope of current `persist` impl.

**Reply by dohomi** [reply]
Ok thanks!
